component/kafka/simple/consumer.go

```diff
@@ -0,0 +1,301 @@
+	err := c.component.proc(btc)
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			return fmt.Errorf("context was cancelled after processing error: %w", err)
+		}
+		if err = c.executeFailureStrategy(messages, err); err != nil {
+			return err
+		}
+	} else {
+		c.mu.Lock()
+		c.updateLatestOffsetReached(messages)
+		c.mu.Unlock()
+	}
+
+	c.mu.Lock()
+	c.hasProcessedMessages = true
+	c.mu.Unlock()
+
+	for _, m := range messages {
+		trace.SpanSuccess(m.Span())
+	}
+
+	return nil
+}
+
+func (c *consumerHandler) getContextWithCorrelation(ctx context.Context, msg *sarama.ConsumerMessage) (context.Context, opentracing.Span) {
+	corID := kafka.GetCorrelationID(msg.Headers)
+
+	sp, ctxCh := trace.ConsumerSpan(ctx, trace.ComponentOpName(consumerComponent, msg.Topic),
+		consumerComponent, corID, mapHeader(msg.Headers))
+	ctxCh = correlation.ContextWithID(ctxCh, corID)
+	ctxCh = log.WithContext(ctxCh, log.Sub(map[string]interface{}{correlation.ID: corID}))
+	return ctxCh, sp
+}
+
+func mapHeader(rh []*sarama.RecordHeader) map[string]string {
+	mp := make(map[string]string, len(rh))
+	for _, h := range rh {
+		mp[string(h.Key)] = string(h.Value)
+	}
+	return mp
+}
+
+func (c *consumerHandler) executeFailureStrategy(messages []kafka.Message, err error) error {
+	switch c.component.failStrategy {
+	case kafka.ExitStrategy:
+		for _, m := range messages {
+			trace.SpanError(m.Span())
+			messageStatusCountInc(messageErrored, m.Message().Partition, m.Message().Topic)
+		}
+		log.Errorf("could not process message(s)")
+		return err
+	case kafka.SkipStrategy:
+		for _, m := range messages {
+			trace.SpanError(m.Span())
+			messageStatusCountInc(messageErrored, m.Message().Partition, m.Message().Topic)
+			messageStatusCountInc(messageSkipped, m.Message().Partition, m.Message().Topic)
+		}
+		c.mu.Lock()
+		c.updateLatestOffsetReached(messages)
+		c.mu.Unlock()
+		log.Errorf("could not process message(s) so skipping with error: %v", err)
+	default:
+		log.Errorf("unknown failure strategy executed")
+		return fmt.Errorf("unknown failure strategy: %v", c.component.failStrategy)
+	}
+	return nil
+}
+
+func (c *consumerHandler) updateLatestOffsetReached(messages []kafka.Message) {
+	if !c.component.notificationOnceReachingLatestOffset() || c.notificationAlreadySent {
+		return
+	}
+
+	// We consume from the end as it will be more efficient to find the latest offset
+	for i := len(messages) - 1; i >= 0; i-- {
+		message := messages[i].Message()
+		partition := message.Partition
+		_, latestOffsetUnreached := c.partitionsWithLatestOffsetUnreached[partition]
+		if !latestOffsetUnreached {
+			continue
+		}
+
+		if c.latestOffsetReached(message, partition) {
+			c.partitionHasReachedLatestOffset(partition)
+			if c.notificationAlreadySent {
+				return
+			}
+		}
+	}
+}
+
+func (c *consumerHandler) latestOffsetReached(message *sarama.ConsumerMessage, partition int32) bool {
+	return message.Offset >= c.component.latestOffsets[partition]
```

Comment:
But the thing is in this call, we need to lock both statements:
https://github.com/teivah/patron/blob/d3010ccc94aaed918aecc46854acb1802d692881/component/kafka/simple/consumer.go#L169-L172
So, if we move the lock inside `updateLatestOffsetReached`, it will be:

```go
func (c *consumerHandler) unit(ctx context.Context, msg *sarama.ConsumerMessage) error {
		// ...
		c.mu.Lock()
		c.hasProcessedMessages = true
		c.mu.Unlock()
		c.updateLatestOffsetReached(messages)
		
		// ...
}

func (c *consumerHandler) updateLatestOffsetReached(messages []kafka.Message) {
		c.mu.Lock()
		defer c.mu.Unlock()
}
```

So we will lock/unlock twice.